core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -305,25 +316,46 @@ private boolean notifySubscriber(Subscriber<Object> subscriber, Queue<Object> qu
 
             if (demand == Long.MAX_VALUE || demandUpdater.compareAndSet(this, demand, demand - 1)) {
                 @SuppressWarnings("unchecked")
-                final T o = (T) queue.remove();
+                T o = (T) queue.remove();
+                ReferenceCountUtil.touch(o);
                 onRemoval(o);
-                if (!subscription.withPooledObjects() && o instanceof ByteBufHttpData) {
-                    ByteBuf buf = ((ByteBufHttpData) o).buf();
-                    try {
-                        subscriber.onNext(HttpData.of(ByteBufUtil.getBytes(buf)));
-                    } finally {
-                        buf.release();
+                if (!subscription.withPooledObjects()) {
+                    if (o instanceof ByteBufHolder) {
```

Comment:
Continuation of https://github.com/line/armeria/pull/595#discussion_r120542867

Do you mind explaining a little more so I don't make the same mistake again? :)

It looks like the old code releases the buf in a finally block so it would be safe. The difference in the new one seems to be support for ByteBuf and not just ByteBufHttpData - do we publish ByteBuf to the stream too? Or is my understanding totally off (quite likely ;) )?